stablehlo/reference/Element.cpp

```diff
@@ -607,6 +601,203 @@ Element ceil(const Element &el) {
   return Element(el.getType(), val);
 }
 
+Element convert(Type type, bool value) {
+  if (isSupportedBooleanType(type)) return Element(type, value);
+  if (isSupportedIntegerType(type))
+    return Element(type, APInt(type.getIntOrFloatBitWidth(), value ? 1 : 0,
+                               type.cast<IntegerType>().isSigned()));
+  if (isSupportedFloatType(type)) {
+    bool roundingErr;
+    auto floatVal = APFloat(value ? 1.0 : 0.0);
+    floatVal.convert(type.cast<FloatType>().getFloatSemantics(),
+                     APFloat::rmNearestTiesToEven, &roundingErr);
+    return Element(type, floatVal);
+  }
+  if (isSupportedComplexType(type)) {
+    bool roundingErr;
+    auto realVal = APFloat(value ? 1.0 : 0.0);
+    realVal.convert(realVal.getSemantics(), APFloat::rmNearestTiesToEven,
+                    &roundingErr);
+    auto imagVal = APFloat(0.0);
+    imagVal.convert(imagVal.getSemantics(), APFloat::rmNearestTiesToEven,
+                    &roundingErr);
+    return Element(type, std::complex<APFloat>(realVal, imagVal));
+  }
+  report_fatal_error(invalidArgument("Unsupported element type: %s",
+                                     debugString(type).c_str()));
+}
+
+Element convert(Type type, int64_t value) {
+  if (isSupportedBooleanType(type)) return Element(type, value != 0);
+  if (isSupportedIntegerType(type))
+    return Element(
+        type, APInt(type.getIntOrFloatBitWidth(), static_cast<uint64_t>(value),
+                    /*isSigned=*/isSupportedSignedIntegerType(type)));
+  if (isSupportedFloatType(type)) {
+    auto floatVal = APFloat(static_cast<double>(value));
+    bool roundingErr;
+    floatVal.convert(type.cast<FloatType>().getFloatSemantics(),
+                     APFloat::rmNearestTiesToEven, &roundingErr);
+    return Element(type, floatVal);
+  }
+  if (isSupportedComplexType(type)) {
+    const llvm::fltSemantics &fltSemantics = type.cast<ComplexType>()
+                                                 .getElementType()
+                                                 .cast<FloatType>()
+                                                 .getFloatSemantics();
+    bool roundingErr;
+    auto realVal = APFloat(static_cast<double>(value));
+    realVal.convert(fltSemantics, APFloat::rmNearestTiesToEven, &roundingErr);
+    auto imagVal = APFloat(0.0);
+    imagVal.convert(fltSemantics, APFloat::rmNearestTiesToEven, &roundingErr);
+    return Element(type, std::complex<APFloat>(realVal, imagVal));
+  }
+  report_fatal_error(invalidArgument("Unsupported element type: %s",
+                                     debugString(type).c_str()));
+}
+
+Element convert(Type type, uint64_t value) {
```

Comment:
The code of the uint64_t function is pretty much the same as the code of the int64_t overload, except that: 1) the int64_t overload has a cast to uint64_t, 2) two lines of code are swapped in the uint64_t overload.

Can we reduce this duplication? 

Also on a related note, I wonder if in the long run we would be better off migrating the interpreter from using APInt to using APSInt. This is what we did in shape refinement recently, and that simplified quite a bit of code - see #1451 for simplifications.

Reply:
> Can we reduce this duplication?
I've deduped other functions, but I'm scratching my head on discerning the signedness of the integer from its primitive type while keeping it as a single function. Plus the change to using `APFloat::convertFromAPInt` made the two functions meaningfully different.

re: APSInt. I also recently discovered this as well last week. Created #1469 to track this.